packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,276 @@
+				 *
+				 * @event sourceEditing
+				 * @param {Object} data Additional information about the event.
+				 * @param {Boolean} data.isSourceEditingMode Flag indicating whether the document source mode is active.
+				 */
+				this.fire( 'sourceEditing', { isSourceEditingMode: this.isSourceEditingMode } );
+			} );
+
+			return buttonView;
+		} );
+
+		// Currently, plugin handles the source editing mode by itself only for the Classic Editor. To use this plugin with other
+		// integrations, listen to `sourceEditing` event and act accordingly.
+		if ( this._isAllowedToHandleSourceEditingMode() ) {
+			this.on( 'sourceEditing', ( evt, { isSourceEditingMode } ) => {
+				if ( isSourceEditingMode ) {
+					this._showSourceEditing();
+					this._disableCommands();
+				} else {
+					this._hideSourceEditing();
+					this._enableCommands();
+				}
+			} );
+
+			this.on( 'change:isEnabled', ( evt, name, isEnabled ) => this._handleReadOnlyMode( !isEnabled ) );
+
+			this.listenTo( editor, 'change:isReadOnly', ( evt, name, isReadOnly ) => this._handleReadOnlyMode( isReadOnly ) );
+		}
+	}
+
+	/**
+	 * Creates source editing wrappers, that replace each editing root. Each wrapper contains the document source from corresponding root.
+	 *
+	 * The wrapper element contains a textarea and it solves the problem, that the textarea element cannot auto expand its height based on
+	 * the content it contains. The solution is to make the textarea more like a plain div element, which expands in height as much as it
+	 * needs to, in order to display the whole document source without scrolling. The wrapper element is a parent for the textarea and for
+	 * the pseudo-element `::after`, that replicates the look, content, and position of the textarea. The pseudo-element replica is hidden,
+	 * but it is styled to be an identical visual copy of the textarea with the same content. Then, the wrapper is a grid container and both
+	 * of its children (the textarea and the `::after` pseudo-element) are positioned within a CSS grid to occupy the same grid cell. The
+	 * content in the pseudo-element `::after` is set in CSS and it stretches the grid to the appropriate size based on the textarea value.
+	 * Since both children occupy the same grid cell, both have always the same height.
+	 *
+	 * @private
+	 */
+	_showSourceEditing() {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+
+		// It is not needed to iterate through all editing roots, as currently the plugin supports only the Classic Editor with single
+		// main root, but this code may help understand and use this feature in external integrations.
+		for ( const [ rootName, domRootElement ] of editingView.domRoots ) {
+			const data = editor.data.get( { rootName } );
+
+			// Remove the current data from the editing root, as the data will be overwritten anyway upon exit from the source editing mode.
+			editor.data.set( { [ rootName ]: '' } );
```

Comment:
What is the actual purpose of this line?

I'm not sure if it is a good idea, it will create another undo step...

Reply:
Yeah, I was wondering how to solve 2 issues:

1. After opening the source editing mode, other buttons in the toolbar become disabled, but they can be "on", if the current selection's attribute match them.
2. In Firefox only, while the table (or other) contextual balloon is displayed and source editing mode is opened, the contextual balloon is re-positioned to the top left edge of the viewport. Sometimes (quite randomly), also the generic Firefox `NS_ERROR_FAILURE` is thrown in this case.

In case of the first issue, it could be quite easily (I think) solve by something like:

```js
model.change( writer => {
    writer.setSelection( null );
    writer.removeSelectionAttribute( model.document.selection.getAttributeKeys() );
} );
```

In case of the second issue, it should be investigated why the balloon is not hidden in Firefox. Quick solution would be to check, if the `ContextualBalloon` is loaded in the editor, and if yes then call `balloon.view.hide()`. I don't know what the consequences will be, to be honest.

Both solutions add some additional logic, and IMHO both issues can be easily solved by removing the current data from the editing root by calling the `editor.data.set()` with an empty string. And important thing is that this one `editor.data.set()` call **should not add undo step** - I hope this will be possible to achieve after the `DataController#set()` method will handle undo.

To sum up, this one line was my first idea to solve both issues at once.

Reply:
After thinking about this problem over the weekend, I changed my solution and now the editor data is no longer deleted.